### PR TITLE
Adjust RPCBus for interface{}-based message passing, and swap the JSON-RPC server for a gRPC server

### LINKED ISSUE #277: Refactor RPCBus to make use of the same `interface{}` based message passing as in the Event Bus

### LINKED ISSUE #278: Incorporate a gRPC server, to facilitate communications across process boundaries

### LINKED ISSUE #280: Make topic encoding independent of the declaration order
The `topics.Topic` is an enum implicitly ordered due to the `iota` syntax. At the moment, the lookup of a topic, given its byte representation, is a simple and unchecked array index access. This means that, upon addition of a topic, one must make sure manually that the number expressed through its byte representation is kept consistent with the index of that topic within the enums. This is error prone and could lead to significant timesinks in case of mistakes. 

A way to enforce and check this consistency at startup would be to introduce an ordering within the `topicBuf` (the _struct_ holding the various topic's byte and string representation) and perform the consistency checks within an `init()` function

### PR DESCRIPTION
**Pull request by julesdesmit**
Fixes #277, fixes #278, fixes #280

### CODE REVIEW

**Comment by julesdesmit** [comment]
Server is currently untested.

**Comment by kieranhall** [comment]
# [Codecov](https://codecov.io/gh/dusk-network/dusk-blockchain/pull/286?src=pr&el=h1) Report
> Merging [#286](https://codecov.io/gh/dusk-network/dusk-blockchain/pull/286?src=pr&el=desc) into [master](https://codecov.io/gh/dusk-network/dusk-blockchain/commit/a2afb1e59db7b87cef6de0e9b36d67a12a1903fa&el=desc) will **increase** coverage by `0.35%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/dusk-network/dusk-blockchain/pull/286?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master     #286      +/-   ##
==========================================
+ Coverage   53.16%   53.51%   +0.35%     
==========================================
  Files         122      121       -1     
  Lines        6939     6901      -38     
==========================================
+ Hits         3689     3693       +4     
+ Misses       2756     2723      -33     
+ Partials      494      485       -9     
```


| [Impacted Files](https://codecov.io/gh/dusk-network/dusk-blockchain/pull/286?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [pkg/core/mempool/mempool.go](https://codecov.io/gh/dusk-network/dusk-blockchain/pull/286/diff?src=pr&el=tree#diff-cGtnL2NvcmUvbWVtcG9vbC9tZW1wb29sLmdv) | `63.56% <0.00%> (+1.35%)` | :arrow_up: |

------

[Continue to review full report at Codecov](https://codecov.io/gh/dusk-network/dusk-blockchain/pull/286?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/dusk-network/dusk-blockchain/pull/286?src=pr&el=footer). Last update [a2afb1e...74dec08](https://codecov.io/gh/dusk-network/dusk-blockchain/pull/286?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

File: pkg/config/groups.go [hunk 1]
```diff
@@ -62,8 +62,8 @@ type rpcConfiguration struct {
 	Network string
 	Address string
 	Enabled bool
-	User    string
```
Comment (ggoranov) [inline pkg/config/groups.go]: If RPC is going to be used over HTTPS (not only UDS), we'll need to secure it with Auth procedure at some point.

File: pkg/rpc/server.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+	rpcBus *rpcbus.RPCBus
+}
+
+// StartgRPCServer starts the gRPC server for the node, to interact with
+// the rust process. It only returns an error, as we want to keep the
+// gRPC service running until the process is killed, thus we do not
+// need to return the server itself.
+func StartgRPCServer(rpcBus *rpcbus.RPCBus) error {
+	// Currently hard-coded on UDS
+	l, err := net.Listen("unix", "/tmp/dusk-grpc.sock")
```
[the remark this answers is not in the document]
Comment (autholykos) [reply pkg/rpc/server.go]: +1

[the remark this answers is not in the document]
Comment (autholykos) [reply pkg/rpc/server.go]: Looks like the address is now handled by the configuration

File: pkg/rpc/server.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	// This function is blocking, so we run it in a goroutine
+	go grpcServer.Serve(l)
+	// TODO: incorporate TLS
+	return nil
+}
+
+func (n *nodeServer) SelectTx(ctx context.Context, req *node.SelectRequest) (*node.SelectResponse, error) {
+	txs, err := n.rpcBus.Call(topics.GetMempoolView, rpcbus.NewRequest(req), 5*time.Second)
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (autholykos) [reply pkg/rpc/server.go]: Good catch. Error logging and handling should definitely be improved.

File: pkg/config/groups.go [hunk 4]
```diff
@@ -62,8 +62,8 @@ type rpcConfiguration struct {
 	Network string
 	Address string
 	Enabled bool
```
[the remark this answers is not in the document]
Comment (autholykos) [reply pkg/config/groups.go]: Yep

File: pkg/config/groups.go (hunk 1 above)
Comment (autholykos) [reply pkg/config/groups.go]: Will create an issue specific to that

**Comment by ggoranov** [comment]
I'm getting a few panics, when closing (CTRL+C) the node connected to a local voucher. Seems the channels are closed and nil value is passed to the handlers.  Not sure if this is reproducible on your side.

@autholykos 

```
INFO[0004] Terminated                                    prefix=main
INFO[0004] Close database                                process=chain
INFO[0004] Handling AutomateConsensusTxs request         prefix=transactor
ERRO[0004] Failed AutomateConsensusTxs request: wallet is not loaded yet  prefix=transactor
panic: interface conversion: interface {} is nil, not bytes.Buffer

goroutine 47 [running]:
github.com/dusk-network/dusk-blockchain/pkg/core/chain.(*Chain).provideRoundResults(0xc00016e410, 0x0, 0x0, 0x0)
        /home/tech/repo/dusknetwork/dusk-blockchain/pkg/core/chain/chain.go:690 +0x63e
github.com/dusk-network/dusk-blockchain/pkg/core/chain.(*Chain).Listen(0xc00016e410)
        /home/tech/repo/dusknetwork/dusk-blockchain/pkg/core/chain/chain.go:168 +0x356
created by main.Setup
        /home/tech/repo/dusknetwork/dusk-blockchain/launch/testnet/server.go:56 +0x2c9
panic: interface conversion: interface {} is nil, not bytes.Buffer

goroutine 48 [running]:
github.com/dusk-network/dusk-blockchain/pkg/core/candidate.(*Broker).provideCandidate(0xc0000dee70, 0x0, 0x0, 0x0)
        /home/tech/repo/dusknetwork/dusk-blockchain/pkg/core/candidate/broker.go:111 +0x479
github.com/dusk-network/dusk-blockchain/pkg/core/candidate.(*Broker).Listen(0xc0000dee70)
        /home/tech/repo/dusknetwork/dusk-blockchain/pkg/core/candidate/broker.go:76 +0x1e2
created by main.Setup
        /home/tech/repo/dusknetwork/dusk-blockchain/launch/testnet/server.go:60 +0x315
panic: interface conversion: interface is nil, not transactions.Transaction

goroutine 35 [running]:
github.com/dusk-network/dusk-blockchain/pkg/core/mempool.Mempool.processSendMempoolTxRequest(0xc000228960, 0xc000228a20, 0xc000228a80, 0xc000228ae0, 0xc000228ba0, 0xe0c9a0, 0xc00021dec0, 0xc000228b40, 0xc0001c0720, 0xffffffff80000000, ...)
        /home/tech/repo/dusknetwork/dusk-blockchain/pkg/core/mempool/mempool.go:484 +0x5c
github.com/dusk-network/dusk-blockchain/pkg/core/mempool.handleRequest(0x0, 0x0, 0x0, 0xc00007bdd8, 0xccc629, 0x6)
        /home/tech/repo/dusknetwork/dusk-blockchain/pkg/core/mempool/mempool.go:548 +0xf1
github.com/dusk-network/dusk-blockchain/pkg/core/mempool.(*Mempool).Run.func1(0xc000162d00)
        /home/tech/repo/dusknetwork/dusk-blockchain/pkg/core/mempool/mempool.go:162 +0x784
created by github.com/dusk-network/dusk-blockchain/pkg/core/mempool.(*Mempool).Run
        /home/tech/repo/dusknetwork/dusk-blockchain/pkg/core/mempool/mempool.go:157 +0x3f
```

**Comment by autholykos** [comment]
I guess this is the reason for Travis failing. Looking into it

**Pull request merged**